### ISSUE TITLE
use weave to allow book pages from `.jmd` and `jl` files.

### ISSUE THREAD

**Issue by rmsrosa** [rmsrosa/booksjl-franklin-template#3]
Following a chat with @TheCedarPrince, who implemented a means of automatically weaving a `.jmd` file to `html` and capture the body to include in the Franklin-generated page, the idea is to adapt this to weave a `jl` script, as well, and hopefully weave them to markdown instead.

From the user's perspective, simply adding a `filename.jl` or `filename.jmd` to the menu should suffice. The methods that build the book layout should figure out the extension and weave the files accordingly.

**Comment by rmsrosa** [comment]
Actually, simply adding `filename.jl` or `filename.jmd` to the menu variable might no work. We will probably need a markdown file `filename.md` containing `{{insert_weave filename.jl}}` or `{{insert_weave filename.jmd}}`, as done in https://github.com/TheCedarPrince/C3T/blob/master/milewski-programmers-1.md.

Hmm, actually, it might work. We may have the function `build_toc`, in `utils.jl`, automatically trigger `insert_weave` when it encounters a filename with `jl` or `jmd` extension.

**Comment by rmsrosa** [comment]
I am having an issue with the figures. Weave sets the figures path relative to out_path, while Franklin moves the contents of say `filename.md` to a new dir `filename/`, which screws the path to the figures. I will try moving the pictures to the proper place after they are generated by weave.

**Comment by tlienart** [comment]
can you specify that in Weave? otherwise it could in Franklin I think

**Comment by rmsrosa** [comment]
Yes, we can specify an output path for the generated markdown file, which can be either an absolute or relative path. I am weaving it to a `_weaved` folder. Say `pages/script.jl` gets weaved to `_weaved/pages/script.md`. We can also specify a figures path, but that is always relative to `out_path`, and the default is `figures`. The generated markdown will refer to figures as `[fig](figures/fig.png)`, which works with the markdown file on the parent folder to `figures`.

Problem is that Franklin renders the markdown in `_weaved/pages/script/index.html` and apparently it ends up looking for the figures in `_weaved/pages/script/figures`, not in `_weaved/pages/figures`. I am trying to move the pictures from one place to the other. It is working locally, when I `serve()` the pages, but the GitHub actions are failing.

I realized I need to check for the existence of the folder before attempting to move. I haven't done that, yet. I had to go out. I will try again later today.

**Comment by rmsrosa** [comment]
I am also pondering whether I should weave it to a gitignored folder or not, say use `__weaved` and make git ignore that.

**Comment by rmsrosa** [comment]
Actually, Franklin should take care of this figure path automatically, right. I don't know why it got it wrong here. Probably I am doing something wrong.

**Comment by tlienart** [comment]
(sorry being a bit lazy) could you prepare a super simple page with a super simple plot that I can look at? and I'll get this path thing fixed :-D 

Note: in the next version (Xranklin) people won't have to deal with path stuff, it will be fully automatic 😄

**Comment by rmsrosa** [comment]
Great! Any expected day for the new version?

The example is there already, but GitHub action is failing so the current gh-page is from a previous version.

There are two plot examples. Here is one: https://github.com/rmsrosa/booksjl-franklin-template/blob/3da12ef6c3cd7c2cf0c393946583e1703794683b/pages/weavejl.jl

**Comment by tlienart** [comment]
I'm a bit confused by how you're trying to do things, did you check the demo? 

https://github.com/tlienart/Franklin.jl/blob/9705c13d39c69ca92d7b37e7453f2394d6e21513/demos/utils.jl#L103-L118

and https://github.com/tlienart/Franklin.jl/blob/master/demos/weave.md ; gives https://franklinjl.org/demos/weave/

this should be adaptable to what you're trying to do and has working figures?

Side note: also why do you want weave support? it doesn't really add anything that Franklin doesn't do apart from the PDF path and it's a heavy dependency, is it just because you happen to have many files in `.jmd` format?

**Comment by rmsrosa** [comment]
Oh, that is exactly what I saw at TheCedarPrince's site. Well, I saw this solution there, but it is for the html output from Weave. I am trying to use the markdown output of Weave, in particular to not have to worry about different CSS's.

Besides, part of my workflow is to write plain julia scripts in VSCode, intertwined with weave-like comments (with lines starting with `#'`). I think it is easier than writing Franklin's type markdown, with fenced codes and worrying about saving and outputting text and figures. These are fine, and convenient, for blog posts or sections where I mainly write text. But when there is a lot of code, just writing julia scripts seems better. That is why I am trying to make it easy to use both.

**Comment by tlienart** [comment]
hmm, I'm not sure I see but in any case you're on your own here; in Franklin you currently have to somehow specify a path for the asset so you'd have to have at some point `\insert_fig{some_path}`; some of this stuff can be lifted to the point where you could have a counter and just do `\insert_latest_fig{}` or `{{last_fig}}` or something of the sorts, but you'll need that input (some indication that there's something to be done at a given point to insert a fig).

That might be fine by you (not sure), if that is, the logic will be a bit annoying because you'll need to keep track of how many cells there has been and make sure your fig insertion happens just after it; this would be easier from inside Franklin but I won't support that work because it's already present in the next version.

The next version is basically a rewrite from scratch which was an enormous amount of work, correspondingly it needs a lot of testing and people trying it before it can be announced; this is why there was a soft announce of the `#xranklin` channel for people who were willing to test it out. ETA is hard to guess; rough plan is to announce a beta version around April time then get more feedback and slowly migrate big sites and maybe, if all goes well, have the version released over the summer. 

I'd understand that this might disturb your timeline so then you're left with either doing this auto-fig yourself (which might be a pain), or dropping this weave stuff for now and focus on the rest (that's what I'd do but then, of course, I don't use Weave 😋 )

**Comment by rmsrosa** [comment]
Hmm, I see, now. The markdown from Weave won't have this path-specifying stuff that is needed for Franklin. Maybe I should stick with the html solution.

Another reason I should have mentioned is that, since this is for a class, I would like to provide either or both a julia script or/and a jupyter notebook of each section, so the students can easily run them and change whatever they want/need. Anyway, I should think over that for a bit.

Or maybe I should give this beta-version of Xranklin a try.

**Comment by tlienart** [comment]
This can be done with Literate: Franklin handles literate more or less seamlessly; have a look at the DataScienceTutorials in JuliaAI, it's all Literate and Jupiter notebooks are generated on the fly.

Ps: yes you can try Xranklin.jl but bearing in mind that it's still early

**Comment by rmsrosa** [comment]
Oh 👀 , I didn't know `Literate.jl` could execute the code! I never looked carefully at the Documentation. I thought the execution was handled separately. So this means I can completely replace `Weave.jl` with `Literate.jl`. And there is even a "Franklin markdown" output flavor, if desired. Nice. I will look into that.

But I should say that I start to get the impression that the problem is not the integration with Weave per se. My current logic is apparently working fine, locally. But github deploy actions are seg faulting with something related to Plots. I switched to Winston and it doesn't seg fault, the gh-action is succesful, but at the end there is a "failed to precompile Winston" message. This should not be related to weave. I don't know. Anyways, I will dig into `Literate.jl`

**Comment by tlienart** [comment]
Getting plots to work well on GithubAction is not easy. This is related to Github, not to Julia.

For Plots to work well, you need to install additional things related to GR. Something like this should work

```
      # GR recommended (plots backend)
      - run: |
             sudo apt-get update -y
             sudo apt-get install -y qt5-default
      
      # Website build
      - run: xvfb-run julia -e ' # here the rest of the Franklin command
```

**Comment by rmsrosa** [comment]
Thanks again! This works!!!

It took me quite some time to get it working, though. I changed a bit the structure and added Plots to a new Project inside `pages/`. Then, within the markdown (a regular Franklin Markdown), I would activate that project and do using Plots, but that didn't work. Deploying didn't fail, but the gh-pages was NOT FOUND. Anyways. I moved Plots back to the main Project, along with Franklin, and now everything seems to work. I am happy with it. And I am happy with Literate, which is what I am using, now. I stopped with Weave. I just need to clean up the mess.

But I am wondering about the strategy you took with DataScienceTutorials.jl. I found the way the menu and stuff is built to be a bit complicate. I want something easier and more automated, and what I have so far is good enough. But I am curious about a different thing. If I am not mistaken, you don't have `__site` in `.gitignore` and what you do is, in fact, build the site locally; push the whole `__site` to github, and then the github action is simply to move the contents of `__site` to gh-pages, without running Franklin in the github action. Is that correct? I realize this speeds up deployment and avoids problems such as this with Plots. How often people do this? Is it common? I am wondering if I should do the same.

**Comment by tlienart** [comment]
> But I am wondering about the strategy you took with DataScienceTutorials.jl. I found the way the menu and stuff is built to be a bit complicate. I want something easier and more automated, and what I have so far is good enough. But I am curious about a different thing. If I am not mistaken, you don't have `__site` in `.gitignore` and what you do is, in fact, build the site locally; push the whole `__site` to github, and then the github action is simply to move the contents of `__site` to gh-pages, without running Franklin in the github action. Is that correct? I realize this speeds up deployment and avoids problems such as this with Plots. How often people do this? Is it common? I am wondering if I should do the same.

Yes, speed only; you'll note that each page has its own environment as well so it takes quite a bit of time to build the whole website from scratch. This could be reduced by making use of the GitHub cache (something that I'm currently doing with Xranklin) this more or less only requires this in the action after the "Git checkout" bit. 

```
      - name: Cache
        uses: actions/cache@v2
        with:
          path: |
                __site
                ~/.julia/artifacts
                ~/.julia/packages
          key: ${{ runner.os }}-franklin-cache-${{ github.sha }}
          restore-keys: ${{ runner.os }}-franklin-cache-
```

In any case if you only push after building (i.e. local -> global) there's no real use of re-building the site on GitHub. It's just convenience. It's not common because people like this convenience of being able to edit a file directly on the GitHub website and get their site rebuilt automatically (e.g. also if someone submits a PR for a typo fix).

Long story short my recommendation would be that, if you can, use a single environment (project.toml) for the whole website and try the cache above if you want to try to speed the deployment up a bit (by avoiding re-building pages that haven't changed).

PS: using the cache can be a bit annoying; so I'd suggest you give it a shot but if you end up wasting time with it, just remove it.

**Comment by rmsrosa** [comment]
I see. I guess I will stick with building at github for the moment, and maybe experiment with the cache thing, but I will worry about this later. I need to do some clean up and close this issue. Many thanks again!

**Comment by rmsrosa** [comment]
Yeah, working fine, now, with both Weave and Franklin. Time to clean up, remove redundancies in the code, and document. It is certainly not over with bugs, but it was a big step. Thanks again, @tlienart !

**Comment by rmsrosa** [comment]
closed by commit 0924618

**Comment by tlienart** [comment]
nice work :)
